qiskit_experiments/calibration/experiments/spectroscopy.py

```diff
@@ -0,0 +1,180 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Spectroscopy experiment class."""
+
+from typing import List, Optional, Tuple, Union
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Gate
+import qiskit.pulse as pulse
+from qiskit.qobj.utils import MeasLevel
+from qiskit_experiments.analysis.curve_fitting import curve_fit
+from qiskit_experiments.base_analysis import BaseAnalysis
+from qiskit_experiments.base_experiment import BaseExperiment
+from qiskit_experiments import AnalysisResult
+from qiskit_experiments.data_processing.data_processor import DataProcessor
+from qiskit_experiments.data_processing.nodes import Probability
+
+
+class SpectroscopyAnalysis(BaseAnalysis):
+    """Class to analysis a spectroscopy experiment."""
+
+    #pylint: disable=arguments-differ, unused-argument
+    def _run_analysis(
+            self,
+            experiment_data,
+            data_processor=None,
+            meas_level=MeasLevel.CLASSIFIED
+    ) -> Tuple[AnalysisResult, None]:
+        """Analyse the spectroscopy experiment. The analysis will loop over the initial
+        guesses of the full width at half maximum.
+        TODO Missing bounds.
+        TODO Missing user-provided intial guess.
+        """
+
+        # Pick a data processor.
+        if data_processor is None:
+            if meas_level == MeasLevel.CLASSIFIED:
+                data_processor = DataProcessor("counts", [Probability("1")])
+            elif meas_level == MeasLevel.KERNELED:
+                data_processor = DataProcessor("memory", [Svd()])
+            else:
+                raise ValueError("Unsupported measurement level.")
+
+        unit = experiment_data.data[0]["metadata"].get("unit", "Hz")
+
+        y_sigmas = np.array([data_processor(datum) for datum in experiment_data.data])
+        sigmas = y_sigmas[:, 1]
+        ydata = abs(y_sigmas[:, 0])
+        xdata = np.array(datum["metadata"]["xval"] for datum in experiment_data.data)
+
+        offset_guess = np.average(ydata)
+        amp_guess = np.max(ydata)
+        peak_idx = np.argmax(ydata)
+        x0_guess = xdata[peak_idx]
+        gamma_guesses = np.linspace(0, abs(xdata[-1] - xdata[0]), 10)
+
+        best_fit = None
+
+        for gamma_guess in gamma_guesses:
+            fit_result = curve_fit(
+                lambda x, a, g, x0, b: a * ( g**2 / ((x-x0)**2 + g**2)) + b,
+                xdata,
+                np.array(ydata),
+                np.array([amp_guess, gamma_guess, x0_guess, offset_guess]),
+                np.array(sigmas)
+            )
+
+            if not best_fit:
+                best_fit = fit_result
+            else:
+                if fit_result["reduced_chisq"] < best_fit["reduced_chisq"]:
+                    best_fit = fit_result
+
```

Comment:
What you do here is general to many experiments. What you say is that the fit is very sensitive to the guess, to the point that one should run multiple fits with different guesses, and select the best fit. I think it's better to modify the `curve_fit` method, such that it agrees to accept a set of guesses, and implement there the loop that runs many fits and selects the best one.

Reply:
@chriseclectic  is this something that you had planned for the curve fit?